1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,248 @@
+
+== What forms can an API Contract take?
+
+Some folks might think "That sounds like documentation", others might think "That sounds like tests". The answer is both, and more.
+
+An API contract can exist in several forms, some more useful than others.
+
+=== The Dreaded Verbal Contract
+
+A common case is the verbal contract, where the API developers discuss it with the frontend or other client application developers as they go. The backend developers write their code, the client developers write theirs, and as various endpoints or properties are conceived, they are explained somehow: literally explained out loud over the conference room table, shouted over a hail of nerf gun pellets and ping pong, or DMed over Slack chats.
+
+----
+Fred: Hey Sarah, there’s a new "fudge" field and it can be "blah" or "whatever"
+
+Sarah: Great! Thanks I’ll chuck that in now.
+----
+
+This works for a while. The API works, the app runs fine, we're making money, and we’re onto the next thing!
+
+This method has a huge number, the main being that communication is hard. You never really know if somebody understood what you meant even if they say they did. Beyond that, not having things written down means people can forget. If another client team also needs the information and they ask another client team, they are getting a copy of a copy which might be slightly wrong.
+
+Another outcome is the original author(s) cannot remember the exact rules, leading to time wasted checking the code. An even worse scenario is that the original author(s) are not reachable for some reason. Maybe the whole team of developers who worked on that API are all on a flight together from New York to Uruguay, and for that entire 9 hour flight nobody can get any answers about how the API is supposed to work, so they're stuck not being able to fix a production issue...
+
+In this instance two common things happen.
+
+1. Developers waste time writing a new endpoint out of fear of using the old one
+2. Developers waste time trying to guess the contract, and might guess wrong
+
+=== Shoddy Human Contracts
+
+When folks first start thinking about contracts, it often takes the form of rudimentary API Reference Documentation. This might be dumped into a Google Doc, shoved in a wiki, or written up in finger blistering HTML. These formats are usually fairly painstaking to create, because you are focusing on formatting _and_ content, shuffling things around whenever you remember that you should list headers too, and copying and pasting examples of JSON which might change over time. API Developers would not settle for writing software like this, but apparently when it comes to documentation it's accepted as the norm.
```

Comment:
```suggestion
When folks first start thinking about contracts, it often takes the form of rudimentary API Reference Documentation. This might be dumped into a Google Doc, shoved in a wiki, or written up in finger blistering HTML. These formats are usually fairly painstaking to create, because you are focusing on formatting _and_ content, shuffling things around whenever you remember that you should list headers too, and copying and pasting examples of JSON which might change over time. API developers would not settle for writing software like this, but apparently when it comes to documentation it's accepted as the norm.
```